For All `CommunityToolkit.Maui.Behavior`s, Ensure `BindingContext` Is Automatically Set to the Attached View's `BindingContext`

brminnick opened the pull request:
### Description of Change ###

The goal of this Pull Request is to ensure that every `Behavior` in `CommunityToolkit.Maui.Behavior` automatically set's the its `BindingContext` to match the `BindingContext` of the attached `View`.

This PR creates/updates the following classes/interfaces:
- `ICommunityToolkitBehavior` - an interface meant for internal consumption, used to align the functionality of `BaseBehavior` and `BasePlatformBehavior`
  ```cs
  protected TView? View { get; set; }
  
  internal bool TrySetBindingContextToAttachedViewBindingContext();
  internal bool TryRemoveBindingContext();
  internal void AssignViewAndBingingContext(TView bindable);
  internal void UnassignViewAndBingingContext(TView bindable);
  protected void OnViewPropertyChanged(object? sender, PropertyChangedEventArgs e)
  protected void OnViewPropertyChanged(TView sender, PropertyChangedEventArgs e);
  ```
- `BaseBehavior : ICommunityToolkitBehavior`
  - `BaseBehavior` now implements `ICommunityToolkitBehavior`
- `BasePlatformBehavior<TView> : BasePlatformBehavior<TView, PlatformView> where TView : Element`
  - Uses the default `PlatformView` for each operating system
- `BasePlatformBehavior<TView, TPlatformView> : PlatformBehavior<TView, TPlatformView>, ICommunityToolkitBehavior<TView> where TView : Element where TPlatformView : class`

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes https://github.com/CommunityToolkit/Maui/discussions/1785

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls <!-- Replace this link to the direct link to your Pull Request in the MicrosoftDocs/CommunityToolkit repo  -->


 ### Additional information ###

The unit tests for `BasePlatformBehavior` can't test `OnAttachedTo()` because `OnAttachedTo()` is only called in platform-specific code for `Microsoft.Maui.Controls.PlatformBehavior`.

I also updated the Unit Tests to use Collection Expressions for a small performance bump.

brminnick commented on src/CommunityToolkit.Maui/Behaviors/BaseBehavior.shared.cs:
```diff
@@ -93,29 +52,17 @@ protected override void OnDetachingFrom(TView bindable)
 	/// <param name="defaultBinding"></param>
 	/// <returns></returns>
-	[MemberNotNullWhen(true, nameof(bindingField), nameof(getContextMethod))]
+	[Obsolete($"{nameof(IsBound)} is no longer used by {nameof(CommunityToolkit)}.{nameof(CommunityToolkit.Maui)} and will be removed in a future release")]
```
I was able to remove the need for this method which removes the need for Reflection.

However, I've kept it here and marked it `[Obsolete]` because removing it would be a breaking change to the API. We can remove this method for the next major version bump.

brminnick commented on src/CommunityToolkit.Maui/Behaviors/ICommunityToolkitBehavior.shared.cs:
```diff
@@ -0,0 +1,96 @@
+/// <summary>
+/// A common interface to be used across <see cref="BaseBehavior{TView}"/> and <see cref="BasePlatformBehavior{TView,TPlatformView}"/>
+/// </summary>
+public interface ICommunityToolkitBehavior<TView> where TView : Element
```
This Interface primarily contains default methods in order to align the functionality of `BaseBehavior` and `BasePlatformBehavior`.

We had to use an interface with default methods because we cannot use the same `abstract class` for both `BaseBehavor : Behavior` and `BasePlatformBehavior : PlatformBehavior` because these base behavior classes inherit from different .NET MAUI classes.

pictos replied:
Should the interface be `internal`?

brminnick replied:
I wish. 

It has to be `public` because all of our public Behaviors are implementing it. 

I did make all of its methods + properties internal or protected so that we don't need to support it as a stand-alone external interface.

pictos commented on src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/BasePlatformBehavior.shared.cs:
```diff
@@ -0,0 +1,75 @@
+	/// </summary>
+	protected TView? View { get; set; }
+
+	TView? ICommunityToolkitBehavior<TView>.View
```
`ICommunityToolkitBehavior` interface is `public`, this will allows any external member to set the `View` value when cast it to the `interface`, is it intentional?

brminnick replied:
It's a `protected` interface property! 

Since it's marked `protected`, devs won't be able to manually set its value for any of our behaviors from a different class. Here's a screenshot of the compiler error I get when trying to change `ICommunityToolkitBehavior.View` in the IDE:

<img width="1133" alt="image" src="https://github.com/CommunityToolkit/Maui/assets/13558917/28106eb4-7fdb-4e3e-b6de-18529b6e8da0">

brminnick replied:
And here's another screenshot from the same code showing that the variable `behavior` is cast to `ICommunityToolkitBehavior`:
<img width="1127" alt="image" src="https://github.com/CommunityToolkit/Maui/assets/13558917/55afe590-91f0-4364-beb2-4c38dbe90889">

pictos commented on src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/BasePlatformBehavior.shared.cs:
```diff
@@ -0,0 +1,75 @@
+	/// <inheritdoc/>
+	protected override void OnAttachedTo(TView bindable, TPlatformView platformView)
+	{
+		base.OnAttachedTo(bindable, platformView);
```
we don't need to call the base method here, [since the implementation is empty](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Interactivity/PlatformBehavior.cs#L128).

brminnick replied:
I prefer to keep the call to `base.OnAttached()` as a best practice. 

If .NET MAUI ever updates their implementation, we'll already be prepared. 

Plus, I literally lost over an hour today trying to figure out why `BasePlatformBehavior.OnAttached()` wasn't getting called. It turns out none of our PlatformBehaviors were calling `base.OnAttached()`. 

Had we always called `base.OnAttached()`, despite it not being immediately necessary, I would've saved a ton of time on this PR.

pictos commented on src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/BasePlatformBehavior.shared.cs:
```diff
@@ -0,0 +1,75 @@
+	protected override void OnDetachedFrom(TView bindable, TPlatformView platformView)
+	{
+		base.OnDetachedFrom(bindable, platformView);
+
```
we don't need to call the base method here, [since the implementation is empty](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Interactivity/PlatformBehavior.cs#L135).

brminnick replied:
I prefer to keep the call to `base.OnAttached()` as a best practice. 

If .NET MAUI ever updates their implementation, we'll already be prepared. 

Plus, I literally lost over an hour today trying to figure out why `BasePlatformBehavior.OnAttached()` wasn't getting called. It turns out none of our PlatformBehaviors were calling `base.OnAttached()`. 

Had we always called `base.OnAttached()`, despite it not being immediately necessary, I would've saved a ton of time on this PR.

pictos commented on src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/IconTintColor/IconTintColorBehavior.macios.cs:
```diff
@@ -119,4 +106,20 @@ static void SetUIImageViewTintColor(UIImageView imageView, View element, Color c
 		imageView.TintColor = color.ToPlatform();
 	}
 
+	void OnElementPropertyChanged(object? sender, PropertyChangedEventArgs e)
```
we already subscribe to the `bindable.PropertyChanged` on base, does make sense to use the override method instead?

brminnick replied:
Great catch! Done  ✅

pictos commented on src/CommunityToolkit.Maui/Behaviors/Validators/MultiValidationBehavior.shared.cs:
```diff
@@ -65,7 +65,7 @@ protected override async ValueTask<bool> ValidateAsync(object? value, Cancellati
 		})).ConfigureAwait(false);
 
-		var errors = children.Where(c => c.IsNotValid).Select(GetError);
+		var errors = children.Where(c => c.IsNotValid).Select(GetError).ToList();
```
It could be a performance regression, so I would ask, do we need to make it a concrete collection here?

From the diff, It's needed when inside the `if`, at line 78. So, it did allocate the collection in a specific case.

brminnick replied:
It's actually a performance bump recommended by reshapes since every path in the code was making multiple calls to the same IEnumerable.

pictos commented on src/CommunityToolkit.Maui/Behaviors/Validators/MultiValidationBehavior.shared.cs:
```diff
@@ -75,31 +75,27 @@ protected override async ValueTask<bool> ValidateAsync(object? value, Cancellati
-					Path = BindingContextProperty.PropertyName,
-					Source = this
-				});
+				((ICommunityToolkitBehavior<VisualElement>)child).TrySetBindingContextToAttachedViewBindingContext();
```
if it fails do we want to log some message or throw an error?

brminnick replied:
Agreed! I'll only use `Trace.WriteLine` when `TrySetBindingContextToAttachedViewBindingContext` fails; I won't use us when `TryRemoveBindingContext` fails since that also returns `false` when there is no `BindingContext` to remove which wouldn't necessarily be an error.

Done ✅

pictos commented on src/CommunityToolkit.Maui/Behaviors/PlatformBehaviors/BasePlatformBehavior.shared.cs:
```diff
@@ -0,0 +1,75 @@
+/// Abstract class for our behaviors to inherit.
+/// </summary>
+/// <typeparam name="TView">The <see cref="VisualElement"/> that the behavior can be applied to</typeparam>
+public abstract partial class BasePlatformBehavior<TView> : BasePlatformBehavior<TView, PlatformView>
```
We need this member to be `public` in order to use on our `public` Behaviors, but do we want external assemblies to create it? As you mentioned on discussion we would like to break the .NET MAUI default behavior outside our lib, and I agree with that. We're setting the `BindingContext` here, but we _know_ what we're doing, could be dangerous to expose it to the world, so my suggestion is to add an `internal ctor` for those `abstract` classes.

brminnick replied:
Done ✅

pictos reviewed (commented):
it looks a lot of work, great that you could fit it soo fast! I've some points there

Axemasta commented:
Does the BaseBehavior need to account for being attached to multiple `VisualElement`s? Currently with the `TouchBehavior` it will get wierd... all views attached will receive inputs but only the last view the behavior was added to will be animated.

Since specifically the `TouchBehavior` needs to be applied 1:1 (behavior : view) should the `BaseBehavior` have a way of preventing attaching to multiple views?

brminnick commented:
> Does the BaseBehavior need to account for being attached to multiple VisualElements? Currently with the TouchBehavior it will get weird... all views attached will receive inputs but only the last view the behavior was added to will be animated.
> 
> Since specifically the TouchBehavior needs to be applied 1:1 (behavior : view) should the BaseBehavior have a way of preventing attaching to multiple views?

Good question - yes! I followed the existing pattern I found in `BaseBehavior` and extended it to `BasePlatformBehavior` via `ICommunityToolkitBehavior`. It'll automatically set the `Behavior.BindingContext` to match the `BindingContext` of the first `VisualElement` it is attached to. If the same `Behavior` is then attached to another `VisualElement`, it will keep the `BindingContext` from the first `VisualElement`.

brminnick commented:
Ok gang! I think this is ready to go. 

I refactored a couple things just now: 
- In `ICommunityToolkitBehavior`, changed `internal OnViewPropertyChanged(object? sender, PropertyChangedEventArgs e)` -> `protected OnViewPropertyChanged(object? sender, PropertyChangedEventArgs e)`
- Implemented CollectionExpressions for Unit Tests

tranb3r commented:
@brminnick @pictos 
I think there is a regression in 8.0.1.
In a CollectionView, the BindingContext is not set correctly when using TouchBehavior. When I tap on an item, the command from another item is executed.
I did not have this problem with 8.0.0 and setting the BindingContext manually.
Also, with 8.0.1, even if I set the BindingContext manually, it does not fix the issue.
Could you please double-check? I do not have the time to build a repro right now, but it's easy to reproduce.
